optimum/bettertransformer/models/decoder_models.py

```diff
@@ -38,8 +38,20 @@ def __init__(self, gpt_layer: "nn.Module", config: "PretrainedConfig"):
         self.gpt_layer = gpt_layer
         self.gpt_layer._attn = self.wrapped_scaled_dot_product
 
-        mask_value = torch.finfo(torch.float32).min
-        self._mask_value = torch.full([], mask_value, dtype=torch.float32)
+        # gpt-2
+        if config.model_type == "gpt2":
+            target_dtype = self.gpt_layer.c_proj.weight.dtype
+        # gpt-neo-x
+        elif config.model_type == "gpt_neox":
+            target_dtype = self.gpt_layer.dense.weight.dtype
+        # gpt-j
+        else:
+            target_dtype = self.gpt_layer.out_proj.weight.dtype
+
+        self.downcast_qk = config.model_type in ["gptj", "gpt_neox"]
+
+        mask_value = torch.finfo(target_dtype).min
+        self._mask_value = torch.full([], mask_value, dtype=target_dtype)
```

Comment:
This will IMO not work because the user may use `model = model.to(torch.float16)` after initializing the model. Here, `self._mask_value` would still be on fp32. I think we really need it in the forward.